web/packages/teleterm/src/services/config/getKeyboardShortcutSchema.ts

```diff
@@ -0,0 +1,176 @@
+/**
+ * Copyright 2023 Gravitational, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { z } from 'zod';
+
+import { Platform } from 'teleterm/mainProcess/types';
+
+const VALID_SHORTCUT_MESSAGE =
+  'A valid shortcut contains zero or more modifiers (like "Shift") and a single key code (like "A" or "Tab"), combined by the "+" character.';
+
+export function invalidKeyCodeIssue(wrongKeyCode: string): z.IssueData {
+  return {
+    code: z.ZodIssueCode.custom,
+    message: `"${wrongKeyCode}" cannot be used as a key code. ${VALID_SHORTCUT_MESSAGE}`,
+  };
+}
+
+export function invalidModifierIssue(wrongModifier: string): z.IssueData {
+  return {
+    code: z.ZodIssueCode.custom,
+    message: `"${wrongModifier}" cannot be used as a modifier. ${VALID_SHORTCUT_MESSAGE}`,
+  };
+}
+
+export function duplicateModifierIssue(): z.IssueData {
+  return {
+    code: z.ZodIssueCode.custom,
+    message: `Duplicate modifier found. ${VALID_SHORTCUT_MESSAGE}`,
+  };
+}
+
+export function getKeyboardShortcutSchema(platform: Platform) {
+  const allowedModifiers = getSupportedModifiers(platform);
+
+  return z
+    .string()
+    .transform(s => s.replaceAll(' ', ''))
+    .transform(s => s.split('+'))
+    .transform(putModifiersFirst(allowedModifiers))
+    .superRefine(validateKeyCodeAndModifiers(allowedModifiers))
+    .transform(s => s.join('+'));
+}
+
+function putModifiersFirst(
+  allowedModifiers: string[]
+): (tokens: string[]) => string[] {
+  return tokens =>
+    tokens.sort((a, b) => {
+      if (allowedModifiers.indexOf(a) === -1) {
+        return 1;
+      }
+      if (allowedModifiers.indexOf(b) === -1) {
+        return -1;
+      }
+      return allowedModifiers.indexOf(a) - allowedModifiers.indexOf(b);
+    });
+}
+
+function validateKeyCodeAndModifiers(
+  allowedModifiers: string[]
+): (tokens: string[], ctx: z.RefinementCtx) => void {
+  return (tokens, ctx) => {
+    // empty accelerator disables the shortcut
+    if (tokens.join('') === '') {
+      return z.NEVER;
+    }
+
+    const [expectedKeyCode, ...expectedModifiers] = [...tokens].reverse();
+
+    if (!ALLOWED_KEY_CODES.includes(expectedKeyCode)) {
+      ctx.addIssue(invalidKeyCodeIssue(expectedKeyCode));
+      return z.NEVER;
+    }
```

Comment:
Thoughts on letting lowercase single characters through as well? I just feel like it doesn't make sense to punish the user for typing in `Commandd+t` rather than `Command+T`. On top of that, when you do provide `Command+t` as the option, the error message you're going to get is:

> keymap.newTab: "t" cannot be used as a key code. A valid shortcut contains zero or more modifiers (like "Shift") and a single key code (like "A" or "Tab"), combined by the "+" character.

Which IMHO doesn't explain well enough that it needs to be uppercase.

Reply:
Ah, it is not that easy, we later use the characters/modifiers to compare event properties or generate tooltips from them.
If we allow "CTRL+SHIFT+PAGEUP" then we won't be able to easily convert it to "Ctrl+Shift+PageUp". 
For now I allowed it only for letters.